File: leetcode/l200.ts
function numIslands(grid: string[][]): number {
    let res = 0;
    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[0].length; j++) {
            if (grid[i][j] === "1") {
                infect(grid, i, j);
                res++;
            }
        }
    }
    return res;
};

function infect(grid: string[][], row: number, col: number) {
    if (row >= 0 && col >= 0 && row < grid.length && col < grid[0].length) {
        if (grid[row][col] === "1") {
            console.log(grid);
            grid[row][col] = "0";
            infect(grid, row - 1, col);
            infect(grid, row, col - 1);
            infect(grid, row + 1, col);
            infect(grid, row, col + 1);
        }
    }
}

File: leetcode/bfs/l200.java
package leetcode.bfs;

public class l200 {
    static int res = 0;
    public int numIslands(char[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        for (int row = 0; row < n; row++) {
            for (int col = 0; col < m; col++) {
                if (grid[row][col] == '1') {
                    res++;
                    bfs(grid, row, col, n, m);
                }
            }
        }
        return res;
    }

    void bfs(char[][] grid,int row,int col,int n,int m) {
        if (grid[row][col] != '1') {
            return;
        }
        grid[row][col] = '2';
        bfs(grid, Math.max(0, row - 1), col, n, m);
        bfs(grid, Math.min(n - 1, row + 1), col, n, m);
        bfs(grid, row, Math.max(0, col - 1), n, m);
        bfs(grid, row, Math.min(m - 1, col + 1), n, m);
    }
}
